Extend GitHub source analysis

ctavan opened the pull request:
I have finally managed to clean up and improve the queries and add some tooling to help everyone reproduce my results.

In contrast to my initial numbers from https://github.com/bcoe/proposal-standard-library-uuid/issues/4 I have changed my analysis in a few ways:

* I am now less strict when collecting the repos to be considered (not only those with top-level package.json files).
* I no longer include repos that use the legacy `node-uuid` module (which happen to be quite a few @broofa: Almost an equal amount of repos I found use `node-uuid` compared to `uuid`…).
* I count each repo where a certain UUID version is being used only once towards the results (before I was stupidly counting lines where `uuid.v1/3/4/5` was appearing which skewed results).
* I calculate a weighted popularity score based on repo watch count.

In addition to the global statistics I have assembled a list of the top 100 GitHub repos that make use of each UUID version in this Google Sheet: https://docs.google.com/spreadsheets/d/1NjrsNgEZaXs10tXBRGgMpA-9rh_a3rEQlKfi1TpAnYI

We may use this sheet in order to extract use cases as suggested in #14 and to potentially get in touch with authors to discuss why they are using `v1` uuids as suggested in https://github.com/bcoe/proposal-standard-library-uuid/issues/4#issuecomment-500823549.

Let me know what you think!

ctavan committed:
Analyze GitHub BigQuery dataset for uuid usage

This adds tooling and documentation around the queries that are used to
get an idea of how the uuid npm module is currently being used in open
source projects.

ctavan commented on data/README.md:
```diff
@@ -1,13 +1,60 @@
-# WIP
```
Not sure about the directory name `data`. Any better suggestions?

broofa replied:
"analysis"? 🤷‍♂

ctavan commented on data/package.json:
```diff
@@ -0,0 +1,11 @@
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "devDependencies": {
```
Should the analysis be a separate node module or should we add these two `devDependencies` to the top level package.json?

broofa commented:
What's the "Seems legit?" column in the table?  (Also, consider renaming that column since putting "No" there implies a project is "not legit", which is probably not a message we want to be sending.)

ctavan commented:
Oh, with "Seems legit?" I meant whether UUID `v1` usage seems a reasonable choice in those projects from my point of view (I didn't mean to question the legitimacy of the repos themselves).

Renamed to "v1 usage reasonable?" and put a "questionable" wherever I believe `v4` would be the better choice and where I intend to get in touch with the maintainers.

Feel free to further improve the wording. Since I'm not a native speaker I sometimes have a hard time finding the right "mood"…

ctavan committed:
Rename analysis directory

ctavan committed:
Exclude flow-typed from analysis to remove false positives

ctavan committed:
Re-run analysis with flow-type excluded

ctavan committed:
Describe hypothesis

ctavan committed:
Mention accidental v1 usage efforts

ctavan committed:
Run prettier on analysis as well

ctavan committed:
Fix analysis link

ctavan committed:
Add FAQ item to explain why v4 is the default

ctavan commented:
@bcoe @codehag I have finally managed to include my hypotheses into the readme of the analysis and I have also added a reference from the main readme to the analysis.

I think this (especially the READMEs) should be good for a review now.

bcoe commented on README.md:
```diff
@@ -95,6 +94,14 @@ How do folks in the community use RFC 4122 UUIDs?
+different versions?**
+
+An analysis of popular Open Source projects that were using `v1` UUIDs has shown, that the majority
+of identified projects was "accidentally" using `v1` UUIDs in situations where `v4` UUIDs would
```
`was accidentally` -> `were accidentally` ... I might instead word it as,

> did not have a compelling reason for using v1, and with education were willing to migrate.

I think we should then plan to reach out to these folks and put some hard numbers around this hypothesis, e.g., we reached out to 10 projects and all but 2 were willing to migrate.

bcoe commented on analysis/README.md:
```diff
@@ -0,0 +1,150 @@
+We want to analyze current usage of the [`uuid` npm module](uuid-npm) in Open Source projects in
+order to design an appropriate API for the UUID standard library.
+
+## Hypothesis
```
this is incredible work.

bcoe commented on analysis/README.md:
```diff
@@ -0,0 +1,150 @@
+in order to reduces the risk of unexpected problems.
+
+- If all you need is a universally unique identifier you should always use `v4` UUIDs.
+- Only if you need time-ordering you should use `v1`.
```
I might flip this sentence and the one below:

> you should only use `v1` if you need time-ordering.

> you should only use `v3`/`v5` if you need namespace based UUID generation.

bcoe commented on analysis/README.md:
```diff
@@ -0,0 +1,150 @@
+
+### Accidental `v1` Usage
+
+**Hypothesis 2:** _A considerable (=more than 50%) amount of `v1` UUID usage is "accidental" in the
```
rather than saying `=more than 50%`, perhaps we can use a confidence interval, .e.g.,

`>50% of `v1` UUID usage is "accidental" with a confidence interval of 95%`

ctavan replied:
OK. But how would you calculate the confidence interval (or rather standard deviation) in our case?

bcoe replied:
My stats are rusty, so we should probably reach out for advice. But, I think, a binomial distribution is what we want. I believe it tells us what range our successes vs., failures fall into.

ctavan replied:
I thought about this one again and I believe we're not on the right track here.

What we essentially want to figure out is whether _enough_ people use `v1` by accident (or because they think that `v1` must somehow be the "default") to justify that we promote `v4` UUIDs in a special defaulty way in our api.

So the question is, what is "enough" to convince us that choosing an interface that promotes `v4` as the default instead of an interface that is symmetric in the 4 versions is a _reasonable_ choice.

I believe specifying an exact threshold and a confidence interval creates pseudo-scientific expectations for a question that just cannot be answered in an exact, quantitative way. Even if we were able to exactly measure the ratio of projects that were using `v1` by accident (which would mean quite some additional effort) and even if we were able to quantify the confidence interval of this result, I doubt that this would help us answer the actual question.

I would argue that it doesn't really matter if 20% of repos use `v1` by accident or maybe even 90%: As long as the amount of accidental `v1` usage is not negligible it would be a strong argument for the asymmetric interface.

The reason why I arbitrarily put a `50%` there was that I assumed nobody would argue that "more than half" could be deemed _negligible_… But as stated above, I think even only 10% or 20% could still be enough…

So how about instead of trying to make this sound more precise (which IMO would be a lie anyways) let's make it sound more fuzzy:

> A non-negligible amount of `v1` UUID usage is "accidental" in the…

And then in the results argue that we consider 8 of the top 10 repos with accidental `v1` usage "non-negligible".

ctavan replied:
@bcoe I've pushed a proposal, let me know what you think.

bcoe replied:
at the end of the day, imagine 50% of people are using `v1` accidentally, all this really does is move us from 89% of people using V4 to 94% of people using V4 (or whatever).

Both numbers are pretty compelling, V4 is a pretty huge majority, and I think we should consider implementing it first (not that we're limiting ourselves from future additions).

ctavan replied:
Please keep in mind that "Hypothesis 2" in my analysis is not so much concerned with the question of whether or not we should support anything but `v4`, that is already covered by "Hypothesis 1".

"Hypothesis 2" is about the question whether we want an interface that nudges users into using `v4` by default and that makes it artificially hard to choose `v1` or `v3`/`v5` "by accident" leaving it to those users who really need the more complex UUID versions (should we ever extend the interface to support those versions as well).

bcoe commented on analysis/analyze.js:
```diff
@@ -0,0 +1,183 @@
+      query: sql,
+      destination,
+      location: 'US'
+      // useQueryCache: false,
```
can we remove these lines?

ctavan replied:
This one yes, I'd like to keep the second comment which is useful when you simply want to repopulate all your tables without removing them before.

I had a hard time finding that option in the BigQuery node client library docs and would love to keep it in the code for reference (it's missing in https://googleapis.dev/nodejs/bigquery/latest/BigQuery.html#createQueryJob and you have to dig into https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobConfigurationQuery to find these options)…

bcoe commented on analysis/analyze.js:
```diff
@@ -0,0 +1,183 @@
+  const availableQueries = queryTemplates.map(query => `  ${query.filename}`).join('\n');
+  const availableQueryIds = ['all'].concat(queryTemplates.map(query => query.id));
+
+  const argv = yargs
```
good library choice 😝

bcoe reviewed (approved):
A few comments, mostly nits. This is incredible work @ctavan.

ctavan committed:
Improve FAQ item on accidental v1 usage

ctavan committed:
Improve background assumptions

ctavan commented:
@bcoe thanks for your review!

Apart from the question regarding standard deviation I have addressed your comments and also rebased onto master.

Feel free to comment again (or merge).

ctavan committed:
Improve hypothesis 2

ctavan committed:
Remove dead code

ctavan commented:
@bcoe as this PR has grown quite big already are you fine with merging it for now and then iterate in future PRs as necessary?

broofa commented:
Any chance we can get an updated analysis for 2026 that includes v6-v8 usage #'s?

FWIW, I slapped [this query](https://gist.github.com/broofa/e77e38e6f58a05c09816197548f46045) together using Claude and it came back with these #'s:

```
version	repo_count	pct_of_repos	total_repos_sampled
v1	12385	71.4	17358
v4	10673	61.5	17358
v5	281	1.6	17358
v3	12	0.1	17358
```

I'm not sure how much I trust this data, though.  I was paying more attention to how to get billing setup than I was the actual query and in hindsight there's some pretty obvious flaws.  (That's what I get for relying on AI I suppose. 🤷 )

... and now I'm waiting to see what the actual cost of that query was before I proceeding.
